tap_mailchimp/streams.py

```diff
@@ -0,0 +1,601 @@
+            self.catalog, self.selected_stream_names, self.child_streams_to_sync)
+        if child_stream_obj.replication_method == 'INCREMENTAL':
+            # Updated the bookmark path with parent id, as the Tap saves bookmark based on parent's id
+            child_stream_obj.bookmark_path[1] = parent_record_id
+        child_stream_obj.path = self.get_path(parent_record_id, child_stream_obj)
+        child_stream_obj.sync()
+
+    def sync(self, sync_start_date=None):
+        """Function to sync records and call child stream for every records"""
+        last_datetime = sync_start_date
+        max_bookmark_field = last_datetime
+
+        ids = []
+        page_size = int(self.config.get('page_size', DEFAULT_PAGE_SIZE))
+        offset = 0
+        has_more = True
+        params = {
+            'count': page_size,
+            **self.params
+        }
+        while has_more:
+            params['offset'] = offset
+
+            # Add param for querying records after a particular date based on the 'bookmark_query_field'
+            if self.bookmark_query_field:
+                params[self.bookmark_query_field] = last_datetime
+
+            LOGGER.info(
+                '%s - Syncing - %scount: %s, offset: %s',
+                self.stream_name,
+                'since: {}, '.format(last_datetime) if self.bookmark_query_field else '',
+                page_size,
+                offset
+            )
+
+            formatted_selected_fields = self.format_selected_fields()
+
+            params['fields'] = formatted_selected_fields
+            data = self.client.get(
+                self.path,
+                params=params,
+                endpoint=self.stream_name)
+
+            raw_records = data.get(self.data_key)
+
+            if len(raw_records) < page_size:
+                has_more = False
+
+            # Loop over every record and sync child stream
+            for record in raw_records:
+                # Remove '_links' as it contains API schema docs
+                del record['_links']
+                # Store 'ids' for reports streams
+                ids.append(record.get('id'))
+                if self.child:
+                    for child in self.child:
+                        # Skip reports child stream sync as it syncs for the list of parent ids
+                        if child in self.report_streams:
+                            continue
+                        # If the child stream is selected or the grandchild is selected then sync the child stream
+                        if child in self.selected_stream_names or child in self.child_streams_to_sync:
+                            self.sync_substream(child, record.get('id'))
+
+            # If the stream is selected then write records
+            if self.to_write_records:
+                max_bookmark_field = self.process_records(raw_records, max_bookmark_field, sync_start_date)
```

Comment:
This implementation looks concerning to me, The parens stream should not be triggering the sync of the child streams, this adds overhead of forcefully including the parent stream in the selected streams, and then working around it to not write the records.

Reply:
> This implementation looks concerning to me, The parens stream should not be triggering the sync of the child streams, this adds overhead of forcefully including the parent stream in the selected streams, and then working around it to not write the records.

https://github.com/singer-io/tap-mailchimp/blob/TDL-20256-dict-to-class-based-implementation/tap_mailchimp/sync.py#L8-L31

this function adds the parent streams to the list of selected streams if any child stream is not selected

Reply:
@leslievandemark  @luandy64 , should we change this impl.?
this will create issue as we further need to modify the behaviour of  child streams.